### ISSUE TITLE
[RFC] Projection implementation

### ISSUE THREAD

**Issue by prolic** [prooph/event-store#234]
Current state is that we have three interfaces (Query, Projection, ReadModelProjection) and the corresponding implementations. Besides that there are some abstractions for those being located in abstract classes and traits.

The problem is, that this might give too many things, users can extend from. It makes it more complicated and might lead to misuse and opened issues because of misunderstandings.

This RFC proposes several solutions to this. Let's discuss the options and find the best fit.

1) Remove all abstractions and have code duplication instead. Then we only have three final classes, where noone can extend from and do unintended things with.

2) Leave some abstractions in traits (they are uselss as is, as you cannot extend them).

3) Make them all anonymous classes inside the event store implementation. You can then only make use of if by calling `$eventStore->createQuery()` and similar calls. No direct instantiation is possible. This makes it completely impossible to extend in a weird way and make false usage. But this means also no abstractions at all are possible, leading to code duplication.

4) like in 3 but we add some traits to reduce code duplication.

5) Leave everything as is.

6) Your suggestions I didn't think of yet.

**Comment by codeliner** [comment]
I'd rephrase 1.:

Remove all abstractions **but keep the interfaces** and have code duplication instead. Then we only have three final classes **(suggested implementation by prooph team) + interfaces if someone want to do it differently**. So noone can extend from and do unintended things with.

**Comment by sandrokeil** [comment]
Agree with @codeliner (we should provide interfaces) but I would also use Traits if it makes sense and userland code can use them.

**Comment by sandrokeil** [comment]
@prolic Do you know or can you imagine if a graph db has any benefits for projections? Maybe we can consider this in the interfaces?

**Comment by prolic** [comment]
Of course we keep the interfaces - it was not up for vote to remove them ;-) Sorry if I was unclear on this. 

About traits: Userland code should not use them at all. A trait would be only used internally to reduce code duplication. We could mark them with `@internal` in phpdoc and comment with don't use these traits on your own, unless you are implementing the event-store interface yourself.

About graph dbs: There are a lot of useful scenarios for them. Currently I don't see what should be added for this. The only relevant methods to create a read model with the ReadModelProjection is by calling

```
public function stack(string $operation, ...$args): void;

// and

public function persist(): void;
```

You can stack whatever operation you want, so the possibilities are endless and there is nothing needed to add, unless you provide an example. See also: https://github.com/prooph/event-store/blob/develop/src/Projection/ReadModel.php

**Comment by codeliner** [comment]
solved by #239
